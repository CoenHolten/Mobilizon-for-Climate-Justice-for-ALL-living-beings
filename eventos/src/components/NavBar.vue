<template>
  <v-toolbar
    class="blue darken-3"
    dark
    app
    clipped-left
    fixed
  >
    <v-toolbar-title style="width: 300px" class="ml-0 pl-3">
      <v-toolbar-side-icon @click.stop="drawer = !drawer"></v-toolbar-side-icon>
      <router-link :to="{ name: 'Home' }">
        Libre-Event
      </router-link>
    </v-toolbar-title>
    <v-select
      autocomplete
      :loading="searchElement.loading"
      light
      solo
      prepend-icon="search"
      placeholder="Search"
      required
      item-text="displayedText"
      :items="searchElement.items"
      :search-input.sync="search"
      v-model="searchSelect"
    ></v-select>
    <v-spacer></v-spacer>
    <v-menu
      offset-y
      :close-on-content-click="false"
      :nudge-width="200"
      v-model="notificationMenu"
    >
      <v-btn icon slot="activator">
        <v-badge left color="red">
          <span slot="badge">{{ notifications.length }}</span>
          <v-icon>notifications</v-icon>
        </v-badge>
      </v-btn>
      <v-card>
        <v-list two-line>
          <template v-for="item in notifications">
            <v-subheader v-if="item.header" v-text="item.header" v-bind:key="item.header"></v-subheader>
            <v-divider v-else-if="item.divider" v-bind:inset="item.inset" v-bind:key="item.inset"></v-divider>
            <v-list-tile avatar v-else v-bind:key="item.title">
              <v-list-tile-content>
                <v-list-tile-title v-html="item.title"></v-list-tile-title>
                <v-list-tile-sub-title v-html="item.subtitle"></v-list-tile-sub-title>
              </v-list-tile-content>
            </v-list-tile>
          </template>
        </v-list>
        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn flat @click="notificationMenu = false">Close</v-btn>
          <v-btn color="primary" flat @click="notificationMenu = false">Save</v-btn>
        </v-card-actions>
      </v-card>
    </v-menu>
    <v-btn flat @click="$router.push({name: 'Account', params: {'id': getUser().account.id}})" v-if="$store.state.user">{{ this.displayed_name }}</v-btn>
  </v-toolbar>
</template>

<script>
  import eventFetch from '@/api/eventFetch';

  export default {
    name: 'NavBar',
    data() {
      return {
        notificationMenu: false,
        notifications: [
          {header: 'Coucou'},
          {title: "T'as une notification", subtitle: 'Et elle est cool'},
        ],
        searchElement: {
          loading: false,
          items: [],
        },
        search: null,
        searchSelect: null,
      };
    },
    watch: {
      search (val) {
        val && this.querySelections(val)
      },
      searchSelect(val) {
        console.log(val);
        if (val.hasOwnProperty('addressLocality')) {
          this.$router.push({name: 'EventList', params: {location: val.geohash}});
        } else {
          this.$router.push({name: 'Account', params: {id: val.id}});
        }
      }
    },
    computed: {
      displayed_name: function() {
        return this.$store.state.user.account.display_name === null ? this.$store.state.user.account.username : this.$store.state.user.account.display_name
      },
    },
    methods: {
      getUser() {
        return this.$store.state.user === undefined ? false : this.$store.state.user;
      },
      querySelections(searchTerm) {
        this.searchElement.loading = true;
        eventFetch('/find/', this.$store, {method: 'POST', body: JSON.stringify({search: searchTerm})})
          .then(response => response.json())
          .then((results) => {
            console.log(results);
            const accountResults = results.accounts.map((result) => {
              if (result.server) {
                result.displayedText = `${result.username}@${result.server.address}`;
              } else {
                result.displayedText = result.username;
              }
              return result;
            });
            const cities = new Set();
            const placeResults = results.places.map((result) => {
              result.displayedText = result.addressLocality;
              return result;
            }).filter((result) => {
              if (cities.has(result.addressLocality)) {
                return false;
              }
              cities.add(result.addressLocality);
              return true;
            });
            this.searchElement.items = accountResults.concat(placeResults);
            this.searchElement.loading = false;
          });
      }
    }
  }
</script>
