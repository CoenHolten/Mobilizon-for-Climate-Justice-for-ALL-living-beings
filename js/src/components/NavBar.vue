<template>
  <nav class="navbar is-fixed-top" role="navigation" aria-label="main navigation">
    <div class="navbar-brand">
      <router-link class="navbar-item" :to="{ name: 'Home' }">Mobilizon</router-link>

      <a
        role="button"
        class="navbar-burger burger"
        aria-label="menu"
        aria-expanded="false"
        data-target="navbarBasicExample"
      >
        <span aria-hidden="true"></span>
        <span aria-hidden="true"></span>
        <span aria-hidden="true"></span>
      </a>
    </div>
    <div class="navbar-end">
      <div class="navbar-item">
        <div class="buttons">
          <router-link class="button is-primary" v-if="!currentUser.id" :to="{ name: 'Register' }">
            <strong>
              <translate>Sign up</translate>
            </strong>
          </router-link>
          <router-link class="button is-light" v-if="!currentUser.id" :to="{ name: 'Login' }">
            <translate>Log in</translate>
          </router-link>
          <router-link
            class="button is-light"
            v-if="currentUser.id && loggedPerson"
            :to="{ name: 'Profile', params: { name: loggedPerson.preferredUsername} }"
          >
            <figure class="image is-24x24">
              <img :src="loggedPerson.avatarUrl">
            </figure>
            <span>{{ loggedPerson.preferredUsername }}</span>
          </router-link>
        </div>
      </div>
    </div>
  </nav>
</template>

<script lang="ts">
import { Component, Vue, Watch } from "vue-property-decorator";
import { SEARCH } from "@/graphql/search";
import { CURRENT_USER_CLIENT } from "@/graphql/user";
import { onLogout } from "@/vue-apollo";
import { deleteUserData } from "@/utils/auth";
import { LOGGED_PERSON } from "@/graphql/actor";
import { IActor, IPerson } from '@/types/actor.model';
import { RouteName } from '@/router'

@Component({
  apollo: {
    search: {
      query: SEARCH,
      variables() {
        return {
          searchText: this.searchText
        };
      },
      skip() {
        return !this.searchText;
      }
    },
    currentUser: {
      query: CURRENT_USER_CLIENT
    },
    loggedPerson: {
      query: LOGGED_PERSON
    }
  }
})
export default class NavBar extends Vue {
  notifications = [
    { header: "Coucou" },
    { title: "T'as une notification", subtitle: "Et elle est cool" }
  ];
  model = null;
  search: any[] = [];
  searchText: string | null = null;
  searchSelect = null;
  loggedPerson!: IPerson;

  get items() {
    return this.search.map(searchEntry => {
      switch (searchEntry.__typename) {
        case "Actor":
          searchEntry.label =
            searchEntry.preferredUsername +
            (searchEntry.domain === null ? "" : `@${searchEntry.domain}`);
          break;
        case "Event":
          searchEntry.label = searchEntry.title;
          break;
      }
      return searchEntry;
    });
  }

  @Watch('model')
  onModelChanged(val) {
    switch (val.__typename) {
      case 'Event':
        this.$router.push({ name: RouteName.EVENT, params: { uuid: val.uuid } });
        break;

      case 'Actor':
        this.$router.push({
          name: RouteName.PROFILE,
          params: { name: this.usernameWithDomain(val) },
        });
        break;
    }
  }

  usernameWithDomain(actor: IActor) {
    return (
      actor.preferredUsername +
      (actor.domain === null ? '' : `@${actor.domain}`)
    );
  }

  enter() {
    console.log("enter");
    this.$apollo.queries["search"].refetch();
  }

  logout() {
    alert('logout !');

    deleteUserData();

    return onLogout(this.$apollo);
  }
}
</script>
