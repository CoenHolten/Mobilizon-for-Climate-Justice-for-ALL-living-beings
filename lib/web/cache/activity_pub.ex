defmodule Mobilizon.Web.Cache.ActivityPub do
  @moduledoc """
  ActivityPub related cache.
  """

  alias Mobilizon.{Actors, Conversations, Events, Resources, Todos, Tombstone}
  alias Mobilizon.Actors.Actor
  alias Mobilizon.Conversations.Comment
  alias Mobilizon.Events.Event
  alias Mobilizon.Federation.ActivityPub.Relay
  alias Mobilizon.Resources.Resource
  alias Mobilizon.Todos.{Todo, TodoList}
  alias Mobilizon.Web.Endpoint
  alias Mobilizon.Web.Router.Helpers, as: Routes

  @cache :activity_pub

  @doc """
  Gets a local actor by username.
  """
  @spec get_local_actor_by_name(String.t()) ::
          {:commit, Actor.t()} | {:ignore, nil}
  def get_local_actor_by_name(name) do
    Cachex.fetch(@cache, "actor_" <> name, fn "actor_" <> name ->
      case Actors.get_local_actor_by_name(name) do
        %Actor{} = actor ->
          {:commit, actor}

        nil ->
          {:ignore, nil}
      end
    end)
  end

  @doc """
  Gets a public event by its UUID, with all associations loaded.
  """
  @spec get_public_event_by_uuid_with_preload(String.t()) ::
          {:commit, Event.t()} | {:ignore, nil}
  def get_public_event_by_uuid_with_preload(uuid) do
    Cachex.fetch(@cache, "event_" <> uuid, fn "event_" <> uuid ->
      case Events.get_public_event_by_uuid_with_preload(uuid) do
        %Event{} = event ->
          {:commit, event}

        nil ->
          with url <- Routes.page_url(Endpoint, :event, uuid),
               %Tombstone{} = tomstone <- Tombstone.find_tombstone(url) do
            tomstone
          else
            _ -> {:ignore, nil}
          end
      end
    end)
  end

  @doc """
  Gets a comment by its UUID, with all associations loaded.
  """
  @spec get_comment_by_uuid_with_preload(String.t()) ::
          {:commit, Comment.t()} | {:ignore, nil}
  def get_comment_by_uuid_with_preload(uuid) do
    Cachex.fetch(@cache, "comment_" <> uuid, fn "comment_" <> uuid ->
      case Conversations.get_comment_from_uuid_with_preload(uuid) do
        %Comment{} = comment ->
          {:commit, comment}

        nil ->
          {:ignore, nil}
      end
    end)
  end

  @doc """
  Gets a resource by its UUID, with all associations loaded.
  """
  @spec get_resource_by_uuid_with_preload(String.t()) ::
          {:commit, Resource.t()} | {:ignore, nil}
  def get_resource_by_uuid_with_preload(uuid) do
    Cachex.fetch(@cache, "resource_" <> uuid, fn "resource_" <> uuid ->
      case Resources.get_resource_with_preloads(uuid) do
        %Resource{} = resource ->
          {:commit, resource}

        nil ->
          {:ignore, nil}
      end
    end)
  end

  @doc """
  Gets a todo list by its UUID, with all associations loaded.
  """
  @spec get_todo_list_by_uuid_with_preload(String.t()) ::
          {:commit, TodoList.t()} | {:ignore, nil}
  def get_todo_list_by_uuid_with_preload(uuid) do
    Cachex.fetch(@cache, "todo_list_" <> uuid, fn "todo_list_" <> uuid ->
      case Todos.get_todo_list(uuid) do
        %TodoList{} = todo_list ->
          {:commit, todo_list}

        nil ->
          {:ignore, nil}
      end
    end)
  end

  @doc """
  Gets a todo by its UUID, with all associations loaded.
  """
  @spec get_todo_by_uuid_with_preload(String.t()) ::
          {:commit, Todo.t()} | {:ignore, nil}
  def get_todo_by_uuid_with_preload(uuid) do
    Cachex.fetch(@cache, "todo_" <> uuid, fn "todo_" <> uuid ->
      case Todos.get_todo(uuid) do
        %Todo{} = todo ->
          {:commit, todo}

        nil ->
          {:ignore, nil}
      end
    end)
  end

  @doc """
  Gets a relay.
  """
  @spec get_relay :: {:commit, Actor.t()} | {:ignore, nil}
  def get_relay do
    Cachex.fetch(@cache, "relay_actor", &Relay.get_actor/0)
  end
end
