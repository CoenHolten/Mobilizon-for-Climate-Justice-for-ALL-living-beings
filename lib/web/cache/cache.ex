defmodule Mobilizon.Web.Cache do
  @moduledoc """
  Facade module which provides access to all cached data.
  """

  alias Mobilizon.Actors.{Actor, Member}
  alias Mobilizon.Conversations.Conversation
  alias Mobilizon.Discussions.{Comment, Discussion}
  alias Mobilizon.Events.Event
  alias Mobilizon.Posts.Post
  alias Mobilizon.Resources.Resource
  alias Mobilizon.Todos.{Todo, TodoList}
  alias Mobilizon.Web.Cache.ActivityPub

  @type local_actor :: %Actor{domain: nil}

  @spec get_actor_by_name(binary) :: {:commit, Actor.t()} | {:ignore, nil}
  defdelegate get_actor_by_name(name), to: ActivityPub
  @spec get_local_actor_by_name(binary) :: {:commit, Actor.t()} | {:ignore, nil}
  defdelegate get_local_actor_by_name(name), to: ActivityPub
  @spec get_public_event_by_uuid_with_preload(binary) :: {:commit, Event.t()} | {:ignore, nil}
  defdelegate get_public_event_by_uuid_with_preload(uuid), to: ActivityPub
  @spec get_comment_by_uuid_with_preload(binary) :: {:commit, Comment.t()} | {:ignore, nil}
  defdelegate get_comment_by_uuid_with_preload(uuid), to: ActivityPub
  @spec get_resource_by_uuid_with_preload(binary) :: {:commit, Resource.t()} | {:ignore, nil}
  defdelegate get_resource_by_uuid_with_preload(uuid), to: ActivityPub
  @spec get_todo_list_by_uuid_with_preload(binary) :: {:commit, TodoList.t()} | {:ignore, nil}
  defdelegate get_todo_list_by_uuid_with_preload(uuid), to: ActivityPub
  @spec get_todo_by_uuid_with_preload(binary) :: {:commit, Todo.t()} | {:ignore, nil}
  defdelegate get_todo_by_uuid_with_preload(uuid), to: ActivityPub

  @spec get_conversation_by_id_with_preload(binary) ::
          {:commit, Conversation.t()} | {:ignore, nil}
  defdelegate get_conversation_by_id_with_preload(uuid), to: ActivityPub
  @spec get_member_by_uuid_with_preload(binary) :: {:commit, Member.t()} | {:ignore, nil}
  defdelegate get_member_by_uuid_with_preload(uuid), to: ActivityPub
  @spec get_post_by_slug_with_preload(binary) :: {:commit, Post.t()} | {:ignore, nil}
  defdelegate get_post_by_slug_with_preload(slug), to: ActivityPub
  @spec get_discussion_by_slug_with_preload(binary) :: {:commit, Discussion.t()} | {:ignore, nil}
  defdelegate get_discussion_by_slug_with_preload(slug), to: ActivityPub
  @spec get_relay :: {:commit, Actor.t()} | {:ignore, nil}
  defdelegate get_relay, to: ActivityPub
end
