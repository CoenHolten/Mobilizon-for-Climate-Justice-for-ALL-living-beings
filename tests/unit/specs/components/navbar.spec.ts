const useRouterMock = vi.fn(() => ({
  push: function () {
    // do nothing
  },
}));

import { shallowMount, VueWrapper } from "@vue/test-utils";
import NavBar from "@/components/NavBar.vue";
import { createMockClient, MockApolloClient } from "mock-apollo-client";
import buildCurrentUserResolver from "@/apollo/user";
import { InMemoryCache } from "@apollo/client/cache";
import { describe, it, vi, expect, afterEach } from "vitest";
import { DefaultApolloClient } from "@vue/apollo-composable";

vi.mock("vue-router/dist/vue-router.mjs", () => ({
  useRouter: useRouterMock,
}));

describe("App component", () => {
  let wrapper: VueWrapper;
  let mockClient: MockApolloClient | null;

  const createComponent = () => {
    const cache = new InMemoryCache({ addTypename: false });

    mockClient = createMockClient({
      cache,
      resolvers: buildCurrentUserResolver(cache),
    });

    wrapper = shallowMount(NavBar, {
      // stubs: ["router-link", "router-view", "o-dropdown", "o-dropdown-item"],
      global: {
        provide: {
          [DefaultApolloClient]: mockClient,
        },
      },
    });
  };

  afterEach(() => {
    wrapper?.unmount();
    mockClient = null;
  });

  it("renders a Vue component", async () => {
    const push = vi.fn();
    useRouterMock.mockImplementationOnce(() => ({
      push,
    }));
    createComponent();

    await wrapper.vm.$nextTick();

    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
    // expect(wrapper.findComponent({ name: "b-navbar" }).exists()).toBeTruthy();
  });
});
