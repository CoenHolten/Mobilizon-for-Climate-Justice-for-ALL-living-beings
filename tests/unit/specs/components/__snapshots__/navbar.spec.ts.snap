// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > renders a Vue component 1`] = `
"<nav class=\\"bg-white border-gray-200 px-2 sm:px-4 py-2.5 dark:bg-zinc-900\\" id=\\"navbar\\">
  <div class=\\"container mx-auto flex flex-wrap items-center mx-auto gap-2 sm:gap-4\\">
    <router-link to=\\"[object Object]\\" class=\\"flex items-center flex-1\\">
      <mobilizon-logo-stub invert=\\"false\\" class=\\"w-40\\"></mobilizon-logo-stub>
    </router-link>
    <!--v-if--><button type=\\"button\\" class=\\"inline-flex items-center p-2 ml-1 text-sm text-zinc-500 rounded-lg md:hidden hover:bg-zinc-100 focus:outline-none focus:ring-2 focus:ring-gray-200 dark:text-zinc-400 dark:hover:bg-zinc-700 dark:focus:ring-gray-600\\" aria-controls=\\"mobile-menu-2\\" aria-expanded=\\"false\\"><span class=\\"sr-only\\">Open main menu</span><svg class=\\"w-6 h-6\\" aria-hidden=\\"true\\" fill=\\"currentColor\\" viewBox=\\"0 0 20 20\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path fill-rule=\\"evenodd\\" d=\\"M3 5a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 15a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z\\" clip-rule=\\"evenodd\\"></path>
      </svg></button>
    <div class=\\"justify-between items-center w-full md:flex md:w-auto md:order-1 hidden\\" id=\\"mobile-menu-2\\">
      <ul class=\\"flex flex-col md:flex-row md:space-x-8 mt-2 md:mt-0 md:font-lightbold\\">
        <!--v-if-->
        <!--v-if-->
        <li>
          <router-link to=\\"[object Object]\\" class=\\"block py-2 pr-4 pl-3 text-zinc-700 border-b border-gray-100 hover:bg-zinc-50 md:hover:bg-transparent md:border-0 md:hover:text-mbz-purple-700 md:p-0 dark:text-zinc-400 md:dark:hover:text-white dark:hover:bg-zinc-700 dark:hover:text-white md:dark:hover:bg-transparent dark:border-gray-700\\">Login</router-link>
        </li>
        <!--v-if-->
      </ul>
    </div>
  </div>
</nav>"
`;
